src/backend/src/lib.rs

```diff
@@ -658,18 +686,116 @@ pub fn get_user_profile() -> Result<UserProfile, GetUserProfileError> {
+/// cycles proportional to the current PoW difficulty.
+///
+/// # Parameters
 ///
-/// Note:
-/// - The chain fusion signer performs threshold key operations including providing public keys,
-///   creating signatures and assisting with performing signed Bitcoin and Ethereum transactions.
+/// * `request` - An optional [`AllowSigningRequest`] that may contain a PoW nonce.
+///   - If [`POW_ENABLED`] is `true` and no request is provided, the call reverts to a
+///     backward-compatible path, calling the signer without the PoW requirement.
+///
+/// # Returns
+///
+/// On success, returns an [`AllowSigningResponse`] that includes:
+/// - The status of the allow-signing request.
+/// - The number of cycles granted for signer operations.
+/// - Information about the completed PoW challenge, if applicable.
 ///
 /// # Errors
-/// Errors are enumerated by: `AllowSigningError`.
+///
+/// This function returns an [`AllowSigningError.PowChallenge`] in cases such as:
+/// - The provided PoW nonce is invalid or missing.
+/// - The user's profile is missing.
+/// - The PoW challenge has expired or was already solved.
+///
+/// # Panics
+///
+/// May panic if internal logic fails or if unexpected conditions occur. (To be expanded with
+/// specific panic conditions as needed.)
 #[update(guard = "may_read_user_data")]
-pub async fn allow_signing() -> Result<(), AllowSigningError> {
-    signer::allow_signing().await
+pub async fn allow_signing(
+    request: Option<AllowSigningRequest>,
+) -> Result<AllowSigningResponse, AllowSigningError> {
+    let principal = ic_cdk::caller();
+
+    // Added for backward-compatibility
+    // TODO remove this code block once the PoW feature has been stabilized
+    if !POW_ENABLED {
+        // Passing None revert to the original cycle calculation logic
+        signer::allow_signing(None).await?;
+        // Propagate errors, otherwise return a placeholder response that frontend can temorarly
+        // ignore.
+        return Ok(AllowSigningResponse {
+            status: AllowSigningStatus::Skipped,
+            allowed_cycles: 0u64,
+            challenge_completion: None,
+        });
+    }
+
+    let request = request.ok_or(AllowSigningError::Other("Invalid request".to_string()))?;
+
+    // The Proof-of-Work (PoW) protection is explicitly enforced at the HTTP entry-point level.
+    // This ensures internal calls to the business service remain unrestricted and do not require
+    // PoW.
+
+    let challenge_completion: ChallengeCompletion = pow::complete_challenge(request.nonce)
+        .map_err(|e| match e {
+            ChallengeCompletionError::InvalidNonce
+            | ChallengeCompletionError::MissingUserProfile
+            | ChallengeCompletionError::ExpiredChallenge
+            | ChallengeCompletionError::MissingChallenge
+            | ChallengeCompletionError::ChallengeAlreadySolved => {
+                AllowSigningError::PowChallenge(e)
+            }
```

Comment:
This can be shortened to:
```suggestion
        .map_err(AllowSigningError::PowChallenge)
```

Reply:
Yes, I will commit your suggestion. The reason I initially added a separate mapping for every ChallengeCompletionError was to make each one explicitly noticeable for anyone reading the code.